contrib/debug-tools/openpaipdb/paipdb.py

```diff
@@ -0,0 +1,216 @@
+        sys.stdout = sys.stdin = handle
+        OCCUPIED.claim(port, sys.stdout)
+
+    def stop_server(self):
+        """Revert the stdin & stdout, close the socket"""
+
+        sys.stdout = self.old_stdout
+        sys.stdin = self.old_stdin
+        OCCUPIED.unclaim(self.port)
+        self.skt.close()
+
+
+def set_trace(addr=DEFAULT_ADDR, port=DEFAULT_PORT, timeout=DEFAULT_TIMEOUT, frame=None):
+    """
+    wrapper function so we can user Paipdb.set_trace() as pdb usage
+    """
+    try:
+        debugger = Paipdb(addr=addr, port=port, timeout=timeout)
+    except socket.error:
+        if OCCUPIED.is_claimed(port, sys.stdout):
+            # paipdb is already on this port just ingnore and go on:
+            print("(Recurrent paipdb invocation ignored)")
+            return
+        else:
+            # Port occupied by something else.
+            raise
+    try:
+        debugger.set_trace(frame or sys._getframe().f_back)
+    except Exception:
+        traceback.print_exc()
+
+
+def _trap_handler(addr, port, timeout, signum, frame):
+    set_trace(addr, port, timeout, frame=frame)
+
+
+def handle_trap(addr=DEFAULT_ADDR, port=DEFAULT_PORT, timeout=DEFAULT_TIMEOUT):
+    """Register rpdb as the SIGTRAP signal handler"""
+    signal.signal(signal.SIGTRAP, partial(_trap_handler, addr, port, timeout))
+
+
+def post_mortem(addr=DEFAULT_ADDR, port=DEFAULT_PORT, timeout=DEFAULT_TIMEOUT):
+    debugger = Paipdb(addr=addr, port=port, timeout=timeout)
+    tb = sys.exc_info()[2]
+    traceback.print_exc()
+    debugger.reset()
+    debugger.interaction(None, tb)
+
+
+# This class is refer to rpbd: https://github.com/tamentis/rpdb/blob/0.1.6/rpdb/__init__.py
+class OccupiedPorts(object):
+    """Maintain paipdb port versus stdin/out file handles.
+
+    Provides the means to determine whether or not a collision binding to a
+    particular port is with an already operating rpdb session.
+
+    Determination is according to whether a file handle is equal to what is
+    registered against the specified port.
+    """
+
+    def __init__(self):
+        self.lock = threading.RLock()
+        self.claims = {}
+
+    def claim(self, port, handle):
+        with self.lock:
+            self.claims[port] = id(handle)
+
+    def is_claimed(self, port, handle):
+        with self.lock:
+            got = (self.claims.get(port) == id(handle))
+        return got
+
+    def unclaim(self, port):
+        with self.lock:
+            del self.claims[port]
+
+# {port: sys.stdout} pairs to track recursive rpdb invocation on same port.
+# This scheme doesn't interfere with recursive invocations on separate ports -
+# useful, eg, for concurrently debugging separate threads.
+OCCUPIED = OccupiedPorts()
+
+def main():
+    args = sys.argv[1:]
+    mainpyfile = args[0]
+    sys.argv[:] = args
+
+    # Do not buffer this message, print to stderr
+    sys.stderr.write("debug host ip is set to {}, port is set to {}, timeout is set to {}\n".format(
```

Comment:
One difference for stdout and stderr is stdout will buffer the message but stderr will not.
So stdout may buffer this message and we can not see this log instantly. 
Since user may want to check this log to get the debug server ip and port, use stderr here.